### DISCUSSION TITLE
Defining boundary conditions on an annular mesh

### DISCUSSION THREAD

**Discussion by amnchacon** [firedrakeproject/firedrake#3352] [Firedrake support]
Drs @dham, @colinjcotter or whoever knows, how can I define the boundary conditions in an annulus mesh? The manual says that "the label on_boundary only refers to the side boundaries that take their labels from the base mesh, and not the top or bottom boundaries." There are other labels to define, but I'm trying and it's not being applied on the boundaries.... and, on the other hand, If I have a cylindrus geometry, how I would define the boundary conditions in this case? Thanks ....

**Comment by sghelichkhani** [comment]
@amnchacon: If you are just asking about the markers for top and bottom boundaries, you should be using `"top"` and `"bottom"` (it says so higher up in the part of the documentation you are citing. )

**Reply by amnchacon** [reply]
Ok. But I'm talking about the annular geometry, so I want to define the boundary conditions in the outer and inner annulus region. Those options ("top" and "buttom") didn't work. Is what I'm talking about clear?

**Reply by dham** [reply]
"top" and "bottom" are the right answer for the boundaries in the extruded direction. Please provide a minimal failing example.

**Reply by amnchacon** [reply]
For example, my mesh is defined like:

**r_outer = 3
r_inner = 2
nr = 5
nt = 30
mesh = AnnulusMesh(r_outer, r_inner, nr, nt)**

And the boundary conditions I tried the following way:

**Tr2 = Constant(75)
Tr1 = Constant(15)
bc1 = DirichletBC(V, Tr2, "top")
bc2 = DirichletBC(V, Tr1, "bottom")**

I don't know if I'm defining them wrongly or if they're ok I'm not seeing where the error is... 

"ValueError: Invalid subdomain 'top': 'top' boundary is identified as 'bottom' boundary in periodic extrusion"

**Reply by ksagiyam** [reply]
`AnnulusMesh` is created by extruding an interval mesh in the circumferential direction, so inner and outer boundaries are `1` and `2`:
```
from firedrake import *


mesh = AnnulusMesh(2., 1., nr=2, nt=8)
V = FunctionSpace(mesh, "CG", 1)
bc1 = DirichletBC(V, 0., 1)
bc2 = DirichletBC(V, 0., 2)
print(bc1.nodes)  # [0 1 2 3 4 5 6 7]
print(bc2.nodes)  # [16 17 18 19 20 21 22 23]
```

Probably, it is better to make a `CircleManifoldMesh` and extrude it radially as:
```
circ = CircleManifoldMesh(8)
mesh = ExtrudedMesh(circ, layers=2, layer_height=.5, extrusion_type="radial")
V = FunctionSpace(mesh, "CG", 1)
bc1 = DirichletBC(V, 0., "top")
bc2 = DirichletBC(V, 0., "bottom")
print(bc1.nodes)  # [ 2  5  8 11 14 17 20 23]
print(bc2.nodes)  # [ 0  3  6  9 12 15 18 21]
```
Please note that outer and inner boundaries are now `"top"` and `"bottom"`.

**Reply by amnchacon** [reply]
Thanks a lot for this information @ksagiyam . I had already tried with the second option and with "top" and "bottom" had worked. Another problem I had was related with with the variational formulation (a specific term). I was not getting to see the temperature contours. Now it's ok. I tried with both type of meshes and all worked succesfully.

Thanks.

**Reply by amnchacon** [reply]
Hello @ksagiyam I have a new question: to define an annular mesh with "AnnulusMesh" function I have to write both of radius (inner and outer), but when I use the second option it is just asked one radius (I imagine is the outer).. and the inner can be defined by specifying the layer_height in the "ExtrudedMesh" function. If my outer radius is 3 and my inner is 2.5 using "AnnulusMesh" function, what should be the layer_height value using the "ExtrudedMesh" function to have equivalent scenarios? what is the general rule for it? 

Thanks @ksagiyam...

**Reply by colinjcotter** [reply]
This is explained here:
https://www.firedrakeproject.org/extruded-meshes.html#id11
